pjmedia/src/pjmedia-codec/and_aud_mediacodec.cpp

```diff
@@ -0,0 +1,1347 @@
+    codec_data = (and_media_private_t*) codec->codec_data;
+
+    /* Create PLC if codec has no internal PLC */
+    if (!and_media_codec[idx].has_native_plc) {
+	pj_status_t status;
+	status = pjmedia_plc_create(pool, and_media_codec[idx].clock_rate,
+				    and_media_codec[idx].samples_per_frame, 0,
+				    &codec_data->plc);
+	if (status != PJ_SUCCESS) {
+	    goto on_error;
+	}
+    }
+
+    /* Create silence detector if codec has no internal VAD */
+    if (!and_media_codec[idx].has_native_vad) {
+	pj_status_t status;
+	status = pjmedia_silence_det_create(pool,
+					and_media_codec[idx].clock_rate,
+					and_media_codec[idx].samples_per_frame,
+					&codec_data->vad);
+	if (status != PJ_SUCCESS) {
+	    goto on_error;
+	}
+    }
+
+    codec_data->pool = pool;
+    codec_data->codec_idx = idx;
+
+    create_codec(codec_data);
+    if (!codec_data->enc || !codec_data->dec) {
+	goto on_error;
+    }
+    pj_mutex_unlock(and_media_factory.mutex);
+
+    *p_codec = codec;
+    return PJ_SUCCESS;
+
+on_error:
+    pj_mutex_unlock(and_media_factory.mutex);
+    and_media_dealloc_codec(factory, codec);
+    return PJMEDIA_CODEC_EFAILED;
+}
+
+/*
+ * Free codec.
+ */
+static pj_status_t and_media_dealloc_codec(pjmedia_codec_factory *factory,
+					   pjmedia_codec *codec )
+{
+    and_media_private_t *codec_data;
+
+    PJ_ASSERT_RETURN(factory && codec, PJ_EINVAL);
+    PJ_ASSERT_RETURN(factory == &and_media_factory.base, PJ_EINVAL);
+
+    /* Close codec, if it's not closed. */
+    codec_data = (and_media_private_t*) codec->codec_data;
+    if (codec_data->enc) {
+        AMediaCodec_stop(codec_data->enc);
+        AMediaCodec_delete(codec_data->enc);
+        codec_data->enc = NULL;
+    }
+
+    if (codec_data->dec) {
+        AMediaCodec_stop(codec_data->dec);
+        AMediaCodec_delete(codec_data->dec);
+        codec_data->dec = NULL;
+    }
+    pj_pool_release(codec_data->pool);
+
+    return PJ_SUCCESS;
+}
+
+/*
+ * Init codec.
+ */
+static pj_status_t and_media_codec_init(pjmedia_codec *codec,
+				        pj_pool_t *pool )
+{
+    PJ_UNUSED_ARG(codec);
+    PJ_UNUSED_ARG(pool);
+    return PJ_SUCCESS;
+}
+
+/*
+ * Open codec.
+ */
+static pj_status_t and_media_codec_open(pjmedia_codec *codec,
+					pjmedia_codec_param *attr)
+{
+    and_media_private_t *codec_data = (and_media_private_t*) codec->codec_data;
+    struct and_media_codec *and_media_data =
+					&and_media_codec[codec_data->codec_idx];
+    unsigned i;
+    pj_status_t status;
+
+    PJ_ASSERT_RETURN(codec && attr, PJ_EINVAL);
+    PJ_ASSERT_RETURN(codec_data != NULL, PJ_EINVALIDOP);
+
+    PJ_LOG(5,(THIS_FILE, "Opening codec.."));
+
+    codec_data->vad_enabled = (attr->setting.vad != 0);
+    codec_data->plc_enabled = (attr->setting.plc != 0);
+    and_media_data->clock_rate = attr->info.clock_rate;
+
+#if PJMEDIA_HAS_AND_MEDIA_AMRNB
+    if (and_media_data->pt == PJMEDIA_RTP_PT_AMR ||
+	and_media_data->pt == PJMEDIA_RTP_PT_AMRWB)
```

Comment:
The usage of enum/hardcoded payload type (e.g: `PJMEDIA_RTP_PT_AMR/WB`) for codec identification should be avoided as much as possible (please check other codec impls, e.g: in ilbc.c PT is used for codec_info.pt field requirement but effectively not really used anywhere else), note that the hardcoded PT may be deprecated in the future, currently in SDP offer generation, the default PT may be changed by PJMEDIA endpoint, and for answerer, it is set to match to PT in the SDP offer. Hardcoded PT may also introduce confusion in debugging code (when the PT are different between SDP & hardcoded).

Reply:
use new callback to predecode

Reply:
Predecode callback does avoid using PT for ID, but still many places using it, like the one here.

My point is, for new codecs like this, using PT (`PJMEDIA_RTP_PT_*` or even keyword `pt/PT`) for identifying codec should be avoided (except to be assigned to `pjmedia_codec_info`), for reasons described above, so please use other identification, simple internal enum should be fine, just avoid `PT`.